crates/runtime/src/datafusion/udf/truncate.rs

```diff
@@ -0,0 +1,722 @@
+                DataFusionError::Internal(format!(
+                    "Failed to downcast to {}",
+                    stringify!($ARRAY_TYPE)
+                ))
+            })?;
+        let width: $SCALAR_TYPE = $WIDTH.try_into().context(WidthCastingFailedSnafu)?;
+        let width_array = <$ARRAY_TYPE>::from_value(width, $ARRAY.len());
+        let result: $ARRAY_TYPE =
+            binary(casted_array, &width_array, |v, w| v - (((v % w) + w) % w))
+                .map_err(|e| DataFusionError::ArrowError(e, None))?;
+        Ok(Arc::new(result) as ArrayRef)
+    }};
+}
+
+fn compute_truncate_scalar(
+    scalar: ScalarValue,
+    width: i64,
+) -> Result<ScalarValue, DataFusionError> {
+    if scalar.is_null() {
+        return Ok(scalar);
+    }
+
+    match scalar {
+        ScalarValue::Int8(Some(v)) => Ok(ScalarValue::Int8(Some(truncate_numeric(v, width)?))),
+        ScalarValue::Int16(Some(v)) => Ok(ScalarValue::Int16(Some(truncate_numeric(v, width)?))),
+        ScalarValue::Int32(Some(v)) => Ok(ScalarValue::Int32(Some(truncate_numeric(v, width)?))),
+        ScalarValue::Int64(Some(v)) => {
+            let result = v - (((v % width) + width) % width);
+            Ok(ScalarValue::Int64(Some(result)))
+        }
+        ScalarValue::UInt8(Some(v)) => Ok(ScalarValue::UInt8(Some(truncate_numeric(v, width)?))),
+        ScalarValue::UInt16(Some(v)) => Ok(ScalarValue::UInt16(Some(truncate_numeric(v, width)?))),
+        ScalarValue::UInt32(Some(v)) => Ok(ScalarValue::UInt32(Some(truncate_numeric(v, width)?))),
+        ScalarValue::UInt64(Some(v)) => Ok(ScalarValue::UInt64(Some(truncate_numeric(v, width)?))),
+        ScalarValue::Decimal128(Some(v), p, s) => {
+            let width = i128::from(width);
+            let result = v - (((v % width) + width) % width);
+            Ok(ScalarValue::Decimal128(Some(result), p, s))
+        }
+        ScalarValue::Decimal256(Some(v), p, s) => {
+            let width = i256::from_i128(i128::from(width));
+            let result = v - (((v % width) + width) % width);
+            Ok(ScalarValue::Decimal256(Some(result), p, s))
+        }
+        ScalarValue::Utf8(Some(mut v)) => {
+            let new_len = usize::try_from(width).context(WidthCastingFailedSnafu)?;
+            v.truncate(new_len);
+            Ok(ScalarValue::Utf8(Some(v)))
+        }
+        ScalarValue::Binary(Some(v)) => {
+            let n = usize::try_from(width).context(WidthCastingFailedSnafu)?;
+            let truncated = v.iter().take(n).copied().collect::<Vec<u8>>();
+            Ok(ScalarValue::Binary(Some(truncated)))
+        }
+        _ => Err(TruncateError::InvalidSecondArgType {
+            data_type: scalar.data_type(),
+        }
+        .into()),
+    }
+}
+
+fn truncate_numeric<V, W>(v: V, w: W) -> Result<V, TruncateError>
+where
+    V: Num + Copy + TryFrom<W, Error = TryFromIntError>,
+{
+    let w = V::try_from(w).context(WidthCastingFailedSnafu)?;
+    Ok(v - (((v % w) + w) % w))
```

Comment:
perhaps refactor the core truncate expression into a function that can be used both here and in the binary array calculation.

Reply:
I know it occurs in several places. I tried for a while to make the truncation logic generic over the Rust primitive types and the Arrow native types with no luck.